src/util/bidi.js

```diff
@@ -198,15 +198,16 @@ let bidiOrdering = (function() {
       order.push(new BidiSpan(0, len - m[0].length, len))
     }
 
-    return order
+    order.forDirection = direction
+    return direction == "rtl" ? order.reverse() : order
```

Comment:
Why not? I could also store a wrapper `{ order: [], direction: ''}` in `line.order`, so that the property is not on an array and doesn't leak.
Global direction changes are probably rare, but this technique would also work for line direction changes, it is simple and lazy, and it exactly implements the logic.